File: Algorithms/0795_Number_of_Subarrays_with_Bounded_Maximum/Python/Number_of_Subarrays_with_Bounded_Maximum_Solution_1.py

# Space: O(1)
# Time: O(n)

class Solution:
    def numSubarrayBoundedMax(self, nums, left, right):
        length = len(nums)
        if length == 1:
            return 1 if left <= nums[0] <= right else 0

        fast, slow, count, res = 0, 0, 0, 0
        while fast < length:
            if left <= nums[fast] <= right:
                count = fast - slow + 1
                res += count
            elif nums[fast] < left:
                res += count
            elif nums[fast] > right:
                slow = fast + 1
                count = 0
            fast += 1
        return res

File: Algorithms/0795_Number_of_Subarrays_with_Bounded_Maximum/Java/NumberofSubarrayswithBoundedMaximumSolution1.java

// Space: O(1)
// Time: O(n)

class Solution {
    public int numSubarrayBoundedMax(int[] nums, int left, int right) {
        if (nums.length == 1) {
            return nums[0] >= left && nums[0] <= right ? 1 : 0;
        }
        int slow = 0, fast = 0, count = 0, res = 0;
        while (fast < nums.length) {
            if (left <= nums[fast] & nums[fast] <= right) {
                count = fast - slow + 1;
                res += count;
            } else if (nums[fast] < left) {
                res += count;
            } else if (right < nums[fast]) {
                slow = fast+1;
                count = 0;
            }
            fast++;
        }
        return res;
    }
}




